resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala

```
@@ -131,6 +133,23 @@ private[spark] class Client(
 
         val podWatcher = new Watcher[Pod] {
           override def eventReceived(action: Action, t: Pod): Unit = {
+            if (action == Action.ADDED) {
+              val ownerRefs = new ArrayBuffer[OwnerReference]
+              ownerRefs += new OwnerReferenceBuilder()
+                .withApiVersion(t.getApiVersion)
+                .withController(true)
+                .withKind(t.getKind)
+                .withName(t.getMetadata.getName)
+                .withUid(t.getMetadata.getUid)
+                .build()
+
+              secret.getMetadata().setOwnerReferences(ownerRefs.asJava)
```

Bugs in the library prevented me from abstracting this out to a common function.


```
        def addOwnerReference(owner: HasMetadata, resource: HasMetadata) = {
          val ownerRefs = new ArrayBuffer[OwnerReference]
          ownerRefs += new OwnerReferenceBuilder()
            .withApiVersion(owner.getApiVersion)
            .withController(true)
            .withKind(owner.getKind)
            .withName(owner.getMetadata.getName)
            .withUid(owner.getMetadata.getUid)
            .build()
          resource.getMetadata().setOwnerReferences(ownerRefs.asJava)
          kubernetesClient.resource(resource).createOrReplace()
        }
```

I couldn't get the above to work for service, although it works for secrets, pods, etc.

---

This seems to compile for me, does it not run properly?

```
            if (action == Action.ADDED) {
              def addOwnerReference(owner: HasMetadata, resource: HasMetadata) = {
                val ownerRefs = new ArrayBuffer[OwnerReference]
                ownerRefs += new OwnerReferenceBuilder()
                  .withApiVersion(owner.getApiVersion)
                  .withController(true)
                  .withKind(owner.getKind)
                  .withName(owner.getMetadata.getName)
                  .withUid(owner.getMetadata.getUid)
                  .build()
                resource.getMetadata().setOwnerReferences(ownerRefs.asJava)
                kubernetesClient.resource(resource).createOrReplace()
              }

              addOwnerReference(t, secret)
              addOwnerReference(t, service)
            }
```

---

Yes, it doesn't work as expected for the service and fails silently without adding any owner references. Ideally, we'd want to PATCH and not PUT but that [doesn't work either](https://github.com/fabric8io/kubernetes-client/issues/635).

---

thanks for the link